clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java

```diff
@@ -0,0 +1,215 @@
+            super(ApiKeys.CREATE_PARTITIONS);
+            this.newPartitions = newPartitions;
+            this.timeout = timeout;
+            this.validateOnly = validateOnly;
+        }
+
+        @Override
+        public CreatePartitionsRequest build(short version) {
+            return new CreatePartitionsRequest(newPartitions, timeout, validateOnly, version);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder bld = new StringBuilder();
+            bld.append("(type=CreatePartitionsRequest").
+                    append(", newPartitions=").append(newPartitions).
+                    append(", timeout=").append(timeout).
+                    append(", validateOnly=").append(validateOnly).
+                    append(")");
+            return bld.toString();
+        }
+    }
+
+    CreatePartitionsRequest(Map<String, NewPartitions> newPartitions, int timeout, boolean validateOnly, short apiVersion) {
+        super(apiVersion);
+        this.newPartitions = newPartitions;
+        this.duplicates = Collections.emptySet();
+        this.timeout = timeout;
+        this.validateOnly = validateOnly;
+    }
+
+    public CreatePartitionsRequest(Struct struct, short apiVersion) {
+        super(apiVersion);
+        Object[] topicCountArray = struct.getArray(TOPIC_PARTITION_COUNT_KEY_NAME);
+        Map<String, NewPartitions> counts = new HashMap<>(topicCountArray.length);
+        Set<String> dupes = Collections.emptySet();
+        for (Object topicPartitionCountObj : topicCountArray) {
+            Struct topicPartitionCountStruct = (Struct) topicPartitionCountObj;
+            String topic = topicPartitionCountStruct.get(TOPIC_NAME);
+            Struct partitionCountStruct = topicPartitionCountStruct.getStruct(NEW_PARTITIONS_KEY_NAME);
+            int count = partitionCountStruct.getInt(COUNT_KEY_NAME);
+            Object[] outerArray = partitionCountStruct.getArray(ASSIGNMENT_KEY_NAME);
+            NewPartitions newPartition;
+            if (outerArray != null) {
+                List<List<Integer>> assignments = new ArrayList(outerArray.length);
+                for (Object inner : outerArray) {
+                    Object[] innerArray = (Object[]) inner;
+                    List<Integer> innerList = new ArrayList<>(innerArray.length);
+                    assignments.add(innerList);
+                    for (Object broker : innerArray) {
+                        innerList.add((Integer) broker);
+                    }
+                }
+                newPartition = NewPartitions.increaseTo(count, assignments);
+            } else {
+                newPartition = NewPartitions.increaseTo(count);
+            }
+            NewPartitions dupe = counts.put(topic, newPartition);
+            if (dupe != null) {
+                if (dupes.isEmpty()) {
+                    dupes = new HashSet<>();
```

Comment:
This isn't worth it, I don't think. Assigning a `new HashSet<>(0)` instead of `emptySet` is probably fine considering all the other work happening in this method.